### PR TITLE
List Outputs for Workspace

### LINKED ISSUE #689: List Outputs
Outputs for a Workspace are now the latest version of all ReleaseFiles for the Workspace and any Snapshots.  The UI will combine these two internal concepts as "Outputs".

However, we have two different types of User viewing them:
* **Unprivileged**: Logged out, or an authenticated User without the `view_release_file` permission.
* **Privileged**: Authenticated and has the `view_release_file` permission.

Privileged users will see all Snapshots (ordered by latest first) and the latest files for the Workspace (first in the list), while unprivileged users will only see the published Snapshots.

Both Snapshots and latest files can redirect to an SPA entry point with relevant API endpoints.

For the privileged user perspective we can add badges for `latest` (the current files), and draft/published status for the Snapshots.

### PR DESCRIPTION
**Pull request by ghickman**
This adds a list page showing the Workspace's current files, and any Snapshots it has.  Users without the `view_release_file` permission will only see publicised Snapshots on the page.

The current files continue to route to the SPA with the Snapshots linking to a new SnapshotDetail view, which is a new SPA entrypoint (with accompanying API endpoint).

**As a user with the correct permissions:**

**As a user without the correct permissions (or logged out):**

Fixes #689

### CODE REVIEW

File: jobserver/templates/workspace_output_list.html
Comment (bloodearnest) [inline jobserver/templates/workspace_output_list.html:60]: This doesn't quite read right to me. I think a date would be better than natural time, and we only need to the day I think, minutes don't matter IMO.

e.g.

View Workspace Outputs:
  * Latest (or Current?)
  * Published on 2021-07-16 by bloodearnest
  * Published on 2021-06-15 by ghickman

File: jobserver/templates/workspace_output_list.html
Comment (bloodearnest) [inline jobserver/templates/workspace_output_list.html:47]: I'd call both of these Outputs or Files, not mix an match. But see below for a more complete suggestion.

File: jobserver/models/outputs.py
Comment (bloodearnest) [inline jobserver/models/outputs.py:127]: Minor nit - I think I prefer `ReleaseSnapshot` more than plan snapshot, as we may have other kinds of snapshots, perhaps. But not a blocker.

**Review by bloodearnest — COMMENTED** [review]
A couple non-blocking minor suggestions, but I think we should resolve the permissions logic question before landing.

File: jobserver/templates/workspace_output_list.html
Comment (bloodearnest) [reply jobserver/templates/workspace_output_list.html:60]: I'm not sure I like putting Latest in the same list as the snapshots. What about:
___________________
| View Current Outputs |
‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
Draft Output Publications:
* 2021-07-16 by bloodearnest (2 files)



Published Outputs:
* Published on  2021-07-16 by bloodearnest (5 files)
* Published on 2021-06-15 by ghickman (2 files)

File: jobserver/views/workspaces.py
Comment (ghickman) [reply jobserver/views/workspaces.py:306]: Definitely agree with this in principle but there are some stricky parts:
1) the exceptions raised are DRF specific and use API specific language
2) the anon check is repeated (for good reason in that function!) before the has_permission function

I'd like to add a few more of these calls, getting the Roles and Permissions we use for gates settled, and then see if we can extract a nice abstraction across all of them.  I think we're still working out some of those bits currently, but we should definitely make sure this is dealt with before the end of the voyage (captured in #699).

File: jobserver/templates/workspace_output_list.html
Comment (ghickman) [reply jobserver/templates/workspace_output_list.html:60]: I really hate writing the text for these parts, so being given better words is hugely appreciated here!

File: jobserver/views/workspaces.py
Comment (bloodearnest) [reply jobserver/views/workspaces.py:306]: Good point.

I think we should probably move into `jobserver/releases.py`, but raise custom logic Exceptions, that the views/DRF can handle as needed.

If `has_permission` checks for anon, then it can be removed for now.

**Review by bloodearnest — APPROVED** [review]

**Comment by ghickman** [comment]
Updated as requested, with a slight tweak to the text on the list page so we don't use "Published" for draft Snapshots:

**As a user with the correct permissions:**


**As a user without the correct permissions (or logged out):**

**Comment by bloodearnest** [comment]
I still think we want to separate out "Current" from the list, to a "View Current Files" button, as it's gonna be the most common path for the user, but not blocking on that, as I'm sure other will have thoughts too.

**Comment by ghickman** [comment]
Agreed on the button, captured in #700

**Pull request merged**
